### DISCUSSION TITLE
ViewModel naviagtion. Do we just use the ViewModel name as THE key?

### DISCUSSION THREAD

**Discussion by munkii** [PrismLibrary/Prism#3135] [General] [answered]
I am guessing that people are sick of the question "How do i do ViewModel first navigation in Prism". I am hoping I can give my answer at the same time as asking this common question and I can just get a nod or a "Yes".

Coming from FreshMVMM to Prism MAUI we have a lot of code that does something like this

`await this.NavigationAsync<SettingsViewModel>();`

I assume we could continue doing this MyViewModel centric navigation by simply registering a custom key of "SettingsViewModel" for Navigation to "Settings" (The  View) like this:

`builder.Services.RegisterForNavigation<Views.Settings.Settings, SettingsViewModel>(nameof(SettingsViewModel));`

**Comment by brianlagunas** [comment]
Correct. You can use the VM name as the navigation URI. To make this even better, you can introduce some extension methods into your project to make registering and navigating easier. Such as `RegisterForViewModelNavigation<Settings, SettingsViewModel>()`, and `NavigatAsync<ViewModel>()`

**Comment by dansiegel** [comment, marked as answer]
As @brianlagunas indicated Prism doesn't directly support this. Navigation is by design key based. Prism uses certain defaults which results in using the View type name if no navigation key is provided. For example by default if you had:

```cs
containerRegistry.RegisterForNavigation<LoginPage, LoginViewModel>();
```

The navigation key would be `LoginPage`. However if you specified a key like:

```cs
containerRegistry.RegisterForNavigation<LoginPage, LoginViewModel("Login");
```

Then the navigation key would be `Login`. However with Prism.Maui's Page navigation specifically you can use the NavigationBuilder to achieve strongly typed ViewModel based navigation like:

```cs
var result = await navigationService.CreateBuilder()
    .AddSegment<LoginViewModel>()
    .NavigateAsync();
```

**Reply by munkii** [reply]
AddSegment was another method that has confused me so far. I was going to ask but thought it was probably a case of RTFM :-) but seeing as you mention it. What is a Segment?
